### PR TITLE
JS Bowling Challenge 

### PR DESCRIPTION
**Pull request by abodian**
User Interface written but ran out of time to test.

### CODE REVIEW

File: lib/bowlingScoreCard.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      this.eachRollScoreArray[rollNumber + 1]
+    );
+  };
+
+  isASpareFrame = (rollNumber) => {
+    return (
+      this.eachRollScoreArray[rollNumber] +
+        this.eachRollScoreArray[rollNumber + 1] ===
+      10
+    );
```
Comment (tmccoy99) [inline lib/bowlingScoreCard.js:43]: I think it could make the code slightly cleaner and more readable if you built some of these methods into a Frame class.

File: lib/bowlingScoreCard.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  roll = (pinDown) => {
+    this.eachRollScoreArray.push(pinDown);
+  };
+
+  overallScore = () => {
+    let score = 0;
+    let rollNumber = 0;
+    for (let i = 0; i < 10; i++) {
+      if (this.isASpareFrame(rollNumber)) {
```
Comment (tmccoy99) [inline lib/bowlingScoreCard.js:14]: Despite the complex use of indices in this method, the clear method names and simple structure allow for code to be easily followed and understood

File: lib/bowlingScoreCard.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        score += this.sumPinsDownInFrame(rollNumber);
+        rollNumber += 2;
+      }
+    }
+    if (this.isASpareFrame(rollNumber - 2)) {
+      score += this.eachRollScoreArray[rollNumber];
+    }
+    return score;
+  };
+
```
Comment (tmccoy99) [inline lib/bowlingScoreCard.js:30]: I understand the means by which you are calculating spare and strike bonuses and it clearly works. However, 
I find that relying on array indexing so heavily will often lead to bugs and complicates building on the code.

File: lib/userInterface.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  start() {
+    console.log(chalk.cyan("Welcome to the Bowling Score Card Application!"));
+    this.playFrame(1, 1);
+  }
+
+  playFrame(frameNum, rollNum) {
+    console.log(chalk.yellow(`Frame ${frameNum} - Roll ${rollNum}`));
+    rl.question("Enter the number of pins down: ", (pinsDown) => {
+      const roll = parseInt(pinsDown);
```
Comment (tmccoy99) [inline lib/userInterface.js:22]: This method could definitely do with being divided into several helper methods, but you are probably aware of this and I understand that your hectic weekend probably didn't allow much time for refactoring.

File: jest/bowlingScoreCard.test.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
+const BowlingScoreCard = require("../lib/bowlingScoreCard");
+
+describe("bowlingScoreCard class", () => {
+  beforeEach(() => {
+    scoreCard = new BowlingScoreCard();
+  });
+
```
Comment (tmccoy99) [inline jest/bowlingScoreCard.test.js:7]: Good use of beforeEach to auto instantiate

File: jest/bowlingScoreCard.test.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    let finalSpare = () => {
+      scoreCard.roll(5);
+      scoreCard.roll(5);
+      scoreCard.roll(10);
+    };
+    oneRolls();
+    finalSpare();
+    expect(scoreCard.overallScore()).toEqual(48);
+  });
+
```
Comment (tmccoy99) [inline jest/bowlingScoreCard.test.js:69]: Your tests are thorough, well-labelled and touch on every edge case. Very impressive.

**Review by tmccoy99 — COMMENTED** [review]
This is very impressive stuff Alex, especially considering the weekend you have had. I think your implementation of the rollNumber variable is smart and shows a good grasp of the algorithm. You also handled all the edge cases, which was no doubt aided by your rigorous testing. My main criticism would be that the code is fairly difficult to both understand and read in places, which would make it more difficult to integrate into a larger project.

File: lib/bowlingScoreCard.js (hunk 1 above)
Comment (abodian) [reply lib/bowlingScoreCard.js:43]: Completely agree with splitting this into a frame class. I had the same thought as I was going along.

File: lib/bowlingScoreCard.js (hunk 3 above)
Comment (abodian) [reply lib/bowlingScoreCard.js:30]: Having looked at your code and use of a linked list, I will be looking into this further as an alternative to array indexing in the future. Thank you

File: lib/userInterface.js (hunk 4 above)
Comment (abodian) [reply lib/userInterface.js:22]: 100% I didn't spend a great deal of time on refactoring, especially with the UI class. Most of my time was spent trying to learn about readline.

File: jest/bowlingScoreCard.test.js (hunk 5 above)
Comment (abodian) [reply jest/bowlingScoreCard.test.js:7]: Makes the test code cleaner, without having to add the new instance of scoreCard manually for each test. :)

File: jest/bowlingScoreCard.test.js (hunk 6 above)
Comment (abodian) [reply jest/bowlingScoreCard.test.js:69]: Thank you
